### PR TITLE
Improve require

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
# Description:

Some changes to make our custom require do less work and to get tests passing against ruby-core.

# Tasks:

- [x] Describe the problem / feature
- [ ] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: test/rubygems/test_require.rb [hunk 1]
```diff
@@ -567,18 +567,20 @@ def util_install_extension_file(name)
...
       io.write <<-C
-        #include <ruby.h>
         void Init_#{name}() { }
       C
     end
 
-    spec.files += ["extconf.rb", "#{name}.c"]
+    write_file File.join(@tempdir, "depend")
+
+    spec.files += ["extconf.rb", "depend", "#{name}.c"]
```
Comment (bronzdoc) [inline test/rubygems/test_require.rb]: why we adding the "depend" string?

Comment (deivid-rodriguez) [reply test/rubygems/test_require.rb]: It's an empty file, actually.

This is related to the compilation of the extension, ruby-core runs against several different systems, and on some systems, an unavailable header was being added to the compilation and causing the extension to not compile. I think this was probably an issue in ruby-core, and I think it's probably fixed by removing the `<ruby.h>` header (I will check), but the idea of the empty `depend` file is that compiling the `.c` file does not depend on any other files, preventing this kind of issues by making the extension "more dummy". I copied it from another dummy extension in some other test.

Anyways, I will review this an follow up.

Comment (deivid-rodriguez) [reply test/rubygems/test_require.rb]: @bronzdoc So if I revert this, compiling the extension tries to pull the `config.h` header which is missing. See for example, this ruby-core test run after I reverted these changes: https://github.com/ruby/ruby/pull/2889/checks?check_run_id=467008623.

I'm not really sure why, but making sure nothing gets pulled with an empty `depend` file fixes it, so I'm setting this PR as ready.

**Comment by bronzdoc** [comment]
@bundlerbot merge

**Comment by deivid-rodriguez** [comment]
Thanks for merging @bronzdoc. After a second thought though, I'd like to revert 4fc0ab21c0f7713829abb522ce3b6d8e24c126b3.

Technically, extensionless ruby files are valid ruby files that can be required. For example, `bin/bundle` is sometimes required from other binstubs even if it's also runnable directly.

So, we should technically consider this kind of files too.

Have a look at #3153.

**Pull request merged**